### PR TITLE
Generate shell completions for bootstrap with Clap

### PR DESCRIPTION
**Pull request by clubby789**
Now that #110693 has been merged, we can look at generating shell completions for x.py with `clap_complete`. Leaving this as draft for now as I'm not sure of the best way to integration the completion generator. Additionally, the generated completions for zsh are completely broken (will need to be resolved upstream, it doesn't seem to handle subcommands + global arguments well).
I don't have Fish installed and would be interested to know how well completions work there.

Alternative to #107827

### CODE REVIEW

**Comment by jyn514** [comment]
Can we support powershell here too?

I saw you're only completing subcommands - now that this is part of bootstrap proper, can we support paths as well?

**Comment by clubby789** [comment]
Paths, as well as the `--` type flags are completed, at least testing under bash. I can add a branch for powershell too since that's supported. As this just hints that arguments may be of type 'file' to bash, completions won't work for, e.g. completing paths from the root while in a subdirectory

**Comment by jyn514** [comment]
Sorry, "paths" is the wrong word - I meant using the `should_run` impls, so that e.g. `x dist l` completes to `x dist llvm-tools`.

**Comment by clubby789** [comment]
Ah, I see - we could potentially do some post-processing on the output but that seems like it could be a bit brittle. It might be possible to tweaks the `Flags` structure a bit so that it can suggest files *or* those paths

**Comment by jyn514** [comment]
:+1: if it it's tricky i'm also happy to land this as it, it already seems like a big improvement :)

**Comment by clubby789** [comment]
Added a new `x run` step for this and check in tidy that they're unchanged. Disabled zsh for now

File: src/bootstrap/run.rs [hunk 1]
```diff
@@ -275,3 +278,34 @@ impl Step for GenerateWindowsSys {
...
+        if let Some(comp) = get_completion(shells::Bash, &bash) {
+            std::fs::write(&bash, comp).expect("writing bash completion");
+        }
+        if let Some(comp) = get_completion(shells::Fish, &fish) {
+            std::fs::write(&fish, comp).expect("writing fish completion");
+        }
+        if let Some(comp) = get_completion(shells::PowerShell, &powershell) {
+            std::fs::write(&powershell, comp).expect("writing powershell completion");
+        }
+        // FIXME: zsh generated completions don't seem to work well with subcommands
```
Comment (jyn514) [inline src/bootstrap/run.rs:301]: Can we report this upstream to `clap_complete`?

File: src/etc/completions/x.sh [hunk 2]
```diff
@@ -0,0 +1,1644 @@
...
+                    COMPREPLY=($(compgen -f "${cur}"))
+                    return 0
+                    ;;
+                --stage)
+                    COMPREPLY=($(compgen -f "${cur}"))
+                    return 0
+                    ;;
+                --keep-stage)
+                    COMPREPLY=($(compgen -f "${cur}"))
+                    return 0
```
Comment (jyn514) [inline src/etc/completions/x.sh:842]: these don't seem right? this will complete a file path for `--stage`.

Comment (clubby789) [reply src/etc/completions/x.sh:842]: Clap defaults to suggesting files. The list of possible hints are [here](https://docs.rs/clap/latest/clap/enum.ValueHint.html) - the best case for this is probably `Other`. I'll fill out the other fields

Comment (jyn514) [reply src/etc/completions/x.sh:842]: if there's a way to make this as an enum of possible options, that would be ideal (maybe with https://docs.rs/clap/latest/clap/struct.Arg.html#method.value_parser ?) but just not adding a suggestion here seems fine too.

Comment (clubby789) [reply src/etc/completions/x.sh:842]: #4896 for `Other`

Comment (clubby789) [reply src/etc/completions/x.sh]: Already merged and a release has been made, so I've updated the dep

File: src/bootstrap/run.rs (hunk 1 above)
Comment (clubby789) [reply src/bootstrap/run.rs:301]: I'll look at properly diagnosing and reporting the issue today

**Review by jyn514 — APPROVED** [review]
this is awesome, thank you!

File: src/bootstrap/run.rs (hunk 1 above)
Comment (clubby789) [reply src/bootstrap/run.rs:301]: https://github.com/clap-rs/clap/issues/4898

Comment (clubby789) [reply src/bootstrap/run.rs]: Even after fixing this error, flags don't seem to be being suggested after the subcommand and the path completion isn't working at all 🤔

**Comment by jyn514** [comment]
@bors r+

This already seems like a big improvement, we can iterate in follow-up PRs :)

Could you also make a PR documenting in this in the dev-guide, maybe around https://rustc-dev-guide.rust-lang.org/building/suggested.html ?

**Comment by clubby789** [comment]
I've just updated the filenames to `x.py.{sh,fish,ps1}`, as this is more conventional (the completions are registed to `x.py`, not `x`).
It may also be useful to duplicate the completions for `x` and `x.py`, but using `x.py` for now as it's the common entrypoint

File: src/bootstrap/run.rs (hunk 1 above)
Comment (clubby789) [reply src/bootstrap/run.rs]: Merged now, I'll investigate the remaining issues with Zsh soon. I'll open a follow-up bumping the clap_complete version when I have a fix

**Pull request merged**